python/ray/_private/worker.py

```diff
@@ -883,6 +882,65 @@ def get_gpu_ids():
     return assigned_ids
 
 
+def get_xpu_ids():
+    """ Get the IDs of the XPUs that are available to the worker.
+
+    If the XPU_VISIBLE_DEVICES environment variable was set when the worker
+    started up, then the IDs returned by this method will be a subset of the
+    IDs in CUDA_VISIBLE_DEVICES. If not, the IDs will fall in the range
+    [0, NUM_GPUS - 1], where NUM_GPUS is the number of GPUs that the node has.
+
+    Returns:
+        A list of XPU IDs
+    """
+    worker = global_worker
+    worker.check_connected()
+
+    if worker.mode != WORKER_MODE:
+        if log_once("worker_get_gpu_ids_empty_from_driver"):
+            logger.warning(
+                "`ray.get_xpu_ids()` will always return the empty list when "
+                "called from the driver. This is because Ray does not manage "
+                "XPU allocations to the driver process."
+            )
+
+    # Get all resources from global core worker
+    all_resource_ids = global_worker.core_worker.resource_ids()
+    assigned_ids = set()
+    for resource, assignment in all_resource_ids.items():
+        # Handle both normal and placement group GPU resources.
+        # Note: We should only get the GPU ids from the placement
+        # group resource that does not contain the bundle index!
+        import re
+
+        if resource == "GPU" or re.match(r"^GPU_group_[0-9A-Za-z]+$", resource):
+            for resource_id, _ in assignment:
+                assigned_ids.add(resource_id)
+
+    assigned_ids = list(assigned_ids)
+    # If the user had already set CUDA_VISIBLE_DEVICES, then respect that (in
+    # the sense that only GPU IDs that appear in CUDA_VISIBLE_DEVICES should be
+    # returned).
+    if global_worker.original_gpu_ids is not None:
+        assigned_ids = [
+            global_worker.original_gpu_ids[gpu_id] for gpu_id in assigned_ids
+        ]
+
+    return assigned_ids
+
+
+@PublicAPI
+@client_mode_hook
+def get_gpu_ids():
+    accelerator = ray._private.utils.get_current_accelerator()
+    ids = []
+    if accelerator == "CUDA":
+        ids = get_cuda_ids()
+    elif accelerator == "XPU":
+        ids = get_xpu_ids()
+    return ids
```

Comment:
What should happen if someone wants to use both CUDA and XPU in their Ray cluster?

Reply:
For most scenarios, NVidia CUDA device and Intel XPU device are unlikely to exist at the same time.
Even if both devices exist at the same time, training or inference impossible to use different devices at the same time, due to lack of a unified computing layer (unified computing layer is used for computing/communicating/transferring on different device type). SYCL is aimed to achieve this unified layer, but actually it is hard.

So, I think it is difficult to use mixing device, and only support to use one accelerator device type.

Reply:
Could we use RAY_EXPERIMENTAL_ACCELERATOR=CUDA/XPU to control what runtime should use? Ray cluster can have them at the same time but use either CUDA or XPU exclusively?

Reply:
This sounds like a viable implementation but it has design tradeoffs.. for example, a Ray task with num_gpus=1 may not work on XPU while working on CUDA. So the user would have to add something to the scheduling request so that it is scheduled onto whatever GPU backend the script was developed on.

Reply:
> This sounds like a viable implementation but it has design tradeoffs.. for example, a Ray task with num_gpus=1 may not work on XPU while working on CUDA. So the user would have to add something to the scheduling request so that it is scheduled onto whatever GPU backend the script was developed on.

Ideally we can do that. but it needs to change the scheduling logic as well. I think it's too much to support this in single PR. Could we support using CUDA or XPU exclusively by RAY_EXPERIMENTAL_ACCELERATOR=CUDA/XPU for now which is the scenario for most users?

Reply:
Let me sync internally with the team. Will get back to you tomorrow.

Reply:
As-is, this PR reduces Ray's ability to manage heterogeneous clusters, which is one of Ray's key features. I can't advocate for merging this with the current design. To fix this, we need support for Ray users to run a cluster with heterogeneous machines.

Take a look at https://github.com/ray-project/ray/pull/37998 / https://github.com/ray-project/enhancements/pull/39 which implements Ray support for Amazon's deep learning accelerators; it includes support for heterogeneous clusters.

Reply:
https://github.com/ray-project/ray/pull/38553/files#diff-bec668cc316d34704cdc3e37dd4ce18e4e87764fdf2c28b1504478c7eb8291deR908
I have closed this PR, and updated all the reviews on PR [https://github.com/ray-project/ray/pull/38553/files](url)

For this heterogeneous clusters review, I have updated the code here [https://github.com/ray-project/ray/pull/38553/files#diff-bec668cc316d34704cdc3e37dd4ce18e4e87764fdf2c28b1504478c7eb8291deR908](url)

Reply:
@cadedaniel thanks for the feedback. we also think #37998 is more flexible, we will rebase our code the take their approach in #38553

Reply:
@cadedaniel I have refactored the implementation like PR #37998 , the code is here #38553, please review it again! Very thanks!